kernel/src/snapshot.rs

```diff
@@ -67,6 +76,115 @@ impl Snapshot {
+        let fs_client = engine.get_file_system_client();
+        let start_version = old_version + 1;
+        // Check for new commits
+        let (new_ascending_commit_files, checkpoint_parts) =
+            log_segment::list_log_files_with_version(
+                fs_client.as_ref(),
+                &log_root,
+                Some(start_version),
+                new_version,
+            )?;
+
+        // NB: we need to check both checkpoints and commits since we filter commits at and below
+        // the checkpoint version. Example: if we have a checkpoint + commit at version 1, the log
+        // listing above will only return the checkpoint and not the commit.
+        if new_ascending_commit_files.is_empty() && checkpoint_parts.is_empty() {
+            match new_version {
+                Some(new_version) if new_version != old_version => {
+                    // No new commits, but we are looking for a new version
+                    return Err(Error::Generic(format!(
+                        "Requested snapshot version {} is newer than the latest version {}",
+                        new_version, old_version
+                    )));
+                }
+                _ => {
+                    // No new commits, just return the same snapshot
+                    return Ok(existing_snapshot.clone());
+                }
+            }
+        }
+
+        // create a log segment just from existing_snapshot.version -> new_version
+        let new_log_segment = LogSegment::try_new(
+            new_ascending_commit_files,
+            checkpoint_parts,
+            log_root.clone(),
+            new_version,
+        )?;
+
+        if new_log_segment.has_checkpoint() {
+            // we have a checkpoint in the new LogSegment, just construct a new snapshot from that
+            return Self::try_new_from_log_segment(
+                existing_snapshot.table_root().clone(),
+                new_log_segment,
+                engine,
+            )
+            .map(Arc::new);
+        }
+
+        // we have new commits and no new checkpoint: we replay new commits for P+M and then
+        // create a new snapshot by combining LogSegments and building a new TableConfiguration
+        let (new_metadata, new_protocol) = new_log_segment.protocol_and_metadata(engine)?;
+        let table_configuration = TableConfiguration::try_new_from(
+            existing_snapshot.table_configuration(),
+            new_metadata,
+            new_protocol,
+            new_log_segment.end_version,
+        )?;
+        // NB: we must add the new log segment to the existing snapshot's log segment
+        let mut ascending_commit_files =
+            existing_snapshot.log_segment.ascending_commit_files.clone();
+        ascending_commit_files.extend(new_log_segment.ascending_commit_files);
+        let combined_log_segment =
+            LogSegment::try_new(ascending_commit_files, vec![], log_root, new_version)?;
```

Comment:
If the old log segment had checkpoint files, don't we need to add them here?

Reply:
I'll add a comment: above we check if there are checkpoint files and return - so once we reach here we know there aren't checkpoints

Reply:
LOL wait good call - the old checkpoint files. fixed now